agent/src/apiclient.rs

```diff
@@ -0,0 +1,296 @@
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Deserialize)]
+pub enum UpdateState {
+    Idle,
+    Available,
+    Staged,
+    Ready,
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Deserialize)]
+#[serde(rename_all = "lowercase")]
+pub enum UpdateCommand {
+    Refresh,
+    Prepare,
+    Activate,
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Deserialize)]
+enum CommandStatus {
+    Success,
+    Failed,
+    Unknown,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct StagedImage {
+    image: Option<UpdateImage>,
+    next_to_boot: bool,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct CommandResult {
+    cmd_type: UpdateCommand,
+    cmd_status: CommandStatus,
+    timestamp: String,
+    exit_status: u32,
+    stderr: String,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct UpdateImage {
+    arch: String,
+    version: String,
+    variant: String,
+}
+
+pub struct OsInfo {
+    pub version_id: String,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct UpdateStatus {
+    update_state: UpdateState,
+    available_updates: Vec<String>,
+    chosen_update: Option<UpdateImage>,
+    active_partition: Option<StagedImage>,
+    staging_partition: Option<StagedImage>,
+    most_recent_command: CommandResult,
+}
+
+pub fn get_raw_args(mut args: Vec<String>) -> Vec<String> {
+    let mut subcommand_args = Vec::new();
+    subcommand_args.push("raw".to_string());
+    subcommand_args.push("-u".to_string());
+    subcommand_args.append(&mut args);
+
+    return subcommand_args;
+}
+
+/// TODO
+/// a helper function to invoke apiclient command
+/// Nit: I'd add some details to this docstring about the fact that the API client is volume mounted into the agent container.
+pub fn invoke_apiclient(args: Vec<String>) -> Result<Output> {
+    let output = Command::new("/bin/apiclient")
+        .args(&args)
+        .output()
+        .context(ApiClientCommand { args })?;
+
+    Ok(output)
+}
+
+pub fn get_update_status() -> Result<UpdateStatus> {
+    let mut raw_args = Vec::new();
+    raw_args.push("/updates/status".to_string());
+
+    let update_status_output = invoke_apiclient(get_raw_args(raw_args))?;
+
+    let update_status_string = String::from_utf8_lossy(&update_status_output.stdout).to_string();
+    let update_status: UpdateStatus =
+        serde_json::from_str(&update_status_string).context(UpdateStatusContent)?;
+
+    Ok(update_status)
+}
+
+/// extract OS info from the local system. For now this is just the version id.
+pub fn get_os_info() -> Result<OsInfo> {
+    let mut raw_args = Vec::new();
+    raw_args.push("/os".to_string());
+
+    let os_info_output = invoke_apiclient(get_raw_args(raw_args))?;
+
+    let os_info_content_string = String::from_utf8_lossy(&os_info_output.stdout).to_string();
+    let os_info_content_json: serde_json::Value =
+        serde_json::from_str(&os_info_content_string).context(OsContent)?;
+
+    Ok(OsInfo {
+        version_id: os_info_content_json["version_id"].to_string(),
+    })
```

Comment:
There should be a way to deserialize the values that you want from the Json string, into the `OsInfo` struct, with serde. Maybe worth looking at how to deserialize with serde:

https://serde.rs/derive.html